internal/config/config.go

```diff
@@ -16,90 +16,183 @@ const (
+	Retries int `mapstructure:"retries"`
 }
 
+// ControlPlane contains config options for the control plane scraper.
 type ControlPlane struct {
-	Enabled           bool                  `mapstructure:"enabled"`
-	ETCD              ControlPlaneComponent `mapstructure:"etcd"`
-	APIServer         ControlPlaneComponent `mapstructure:"apiServer"`
+	// Enabled controls whether control plane scraping will be attempted, for any component.
+	Enabled bool `mapstructure:"enabled"`
+	// ETCD contains configuration for the etcd scraper.
+	ETCD ControlPlaneComponent `mapstructure:"etcd"`
+	// APIServer contains configuration for the API server scraper.
+	APIServer ControlPlaneComponent `mapstructure:"apiServer"`
+	// ControllerManager contains configuration for the controller manager scraper.
 	ControllerManager ControlPlaneComponent `mapstructure:"controllerManager"`
-	Scheduler         ControlPlaneComponent `mapstructure:"scheduler"`
-	Timeout           time.Duration         `mapstructure:"timeout"`
-	Retries           int                   `mapstructure:"retries"`
+	// Scheduler contains configuration for the scheduler scraper.
+	Scheduler ControlPlaneComponent `mapstructure:"scheduler"`
+	// Timeout controls the timeout for the requests to control plane endpoints.
+	Timeout time.Duration `mapstructure:"timeout"`
+	// Retries controls how many times the integration will attempt to connect to control plane components before giving up.
+	Retries int `mapstructure:"retries"`
 }
 
+// ControlPlaneComponent contains the config for a control plane component.
 type ControlPlaneComponent struct {
-	Enabled        bool                       `mapstructure:"enabled"`
-	StaticEndpoint *Endpoint                  `mapstructure:"staticEndpoint"`
-	Autodiscover   []AutodiscoverControlPlane `mapstructure:"autodiscover"`
+	// Enabled controls whether this particular component should be scraped.
+	Enabled bool `mapstructure:"enabled"`
+	// StaticEndpoint contains an Endpoint configuration. If set, Autodiscover will not be attempted and the integration
+	// will contact this endpoint directly instead.
+	// Please note that failure to connect to a StaticEndpoint is considered a fatal error and will cause the
+	// integration to exit with a non-zero code.
+	StaticEndpoint *Endpoint `mapstructure:"staticEndpoint"`
+	// Autodiscover contains one or more criteria for discovering control plane endpoints. The first criteria that
+	// matches an endpoint will be attempted.
+	// If an endpoint matching the criteria is discovered but connection fails, the integration will skip to the next
+	// endpoint.
+	// The list of discovered endpoints is exhausted, either because of failure or because it is empty, the integration
+	// will understand there is no control plane component discoverable _at the current point in time_, and will keep
+	// probing for endpoints periodically.
+	Autodiscover []AutodiscoverControlPlane `mapstructure:"autodiscover"`
```

Comment:
@gsanchezgavier Can you take a look at this statements and ensure they are correct? I wrote this from the top of my head and I do remember that we changed the approach a couple times here.

Reply:
this is correct but i would clarify the difference between the Pod discovery and the endpoint probing. I made a suggestion below , feel free to take something or modify it

> The component will iterate over the Autodiscover entries looking for the first Pod that matches the criteria  (selector, namespace and matchNode). Then it probes the listed endpoints for that Autodiscover until the first one successfully responding.
If no Pod is discovered or non of the listed endpoints probes succeeds the he integration will understand there is no control plane component discoverable at the current point in time, and will keep probing.

Reply:
I'm still missing some bits, I think. Let me write a series of use cases here and we can check if they are correct:
| Scenario | Result |
| ---------|---------|
| The first autodiscovery entry does not find a matching pod | The integration will try the next autodiscovery entry |
| None of the autodiscovery entries find a matching pod | The integration will keep probing for pods in case they appear later |
| All the endpoints configured for the first autodiscovery entry, which did find a pod, fail | ?? |

Reply:
for that last one it fails https://github.com/newrelic/nri-kubernetes/blob/main/src/controlplane/scraper.go#L255

Reply:
so my explanation was wrong I will edit that comment

Reply:
🤦 i'm wrong , i was right at the begging , this is where that error is handled . sorry for the mess 
https://github.com/newrelic/nri-kubernetes/blob/main/src/controlplane/scraper.go#L264

Reply:
I'm still a bit confused here 😅 
One more relevant point: https://github.com/newrelic/nri-kubernetes/blob/1e3231333f92f311afc7d1bcd2fbbb3855670293/src/controlplane/discoverer/discoverer.go#L92-L92

Reply:
@gsanchezgavier So what would be the behavior for the third scenario in the table?

Reply:
in that case it will not fail but it will not try with the rest of autodiscovery entries.

Reply:
That's a bit weird, we might want to revisit that. I'll document it for now until we change it.

Reply:
sure we can review it, not sure if there is a case where multiple entires could match in a cluster

Reply:
Updated the comment accordingly, please LMK your thoughts @gsanchezgavier :)